### PR TITLE
Add linkcheck

### LINKED ISSUE #439: Add broken link check to CI for documentation builds
## 📚 Documentation
cc @kne42 

Proposal: add a linkchecker step to our [CI for documentation builds](https://github.com/napari/napari/blob/master/.github/workflows/make_docs.yml). The docs build should fail if it finds broken links.

I think this python package would do the job well, it's super simple to use: https://github.com/linkchecker/linkchecker

Reason for change: there are five broken links right now on the [napari.org](https://napari.org) main page alone. Lots more on other pages too. (Often we see this because the structure of the napari repo has changed, or a file is renamed slightly. There's less evidence of link rot from external sites we point to.)

### PR DESCRIPTION
**Pull request by laysauchoa**
Add linkcheck to spot broken links

Signed-off-by: laysa.uchoa <laysa.uchoa@gmail.com>

# Description

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by laysauchoa** [comment]
Can someone help me to run the workflows? due to: **"First-time contributors need a maintainer to approve running workflows. [Learn more.](https://docs.github.com/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks) 3 successful checks"** Thanks ^^

**Comment by Czaki** [comment]
Workflow run

**Comment by Czaki** [comment]
The main problem (different from broken links) that I see here is that we build docs twice. 

Some observations about broken links. 

@ziyangczi is now @potating-potato on github. 

dask webpage looks broken

There are some broken links to Wikipedia with a lack of closing `)` this is problem with markdown syntax. https://stackoverflow.com/questions/13824669/how-do-you-write-a-link-containing-a-closing-bracket-in-markdown-syntax

File: Makefile [hunk 1]
```diff
@@ -49,3 +49,6 @@ watch:
 			--signal SIGKILL \
 			napari -- $(WATCH_ARGS) || \
 		echo "please run 'pip install watchdog[watchmedo]'"
+
+linkcheck-files:
+	NAPARI_APPLICATION_IPY_INTERACTIVE=0 sphinx-build -b linkcheck --color docs/ docs/_build
```
Comment (alisterburt) [inline Makefile]:
```suggestion
	NAPARI_APPLICATION_IPY_INTERACTIVE=0 sphinx-build -b linkcheck --color docs/ docs/_build

```

**Review by alisterburt — COMMENTED** [review]
This seems to work great! Thank you so much for the fast PR

@Czaki and @laysauchoa do you think there is a way to avoid building the docs twice? It's not obvious to me, unless we had some mechanism to just resolve all links in the docs in advance.

If there isn't an easy way to avoid building the docs twice I vote we push forwards on fixing links - @Czaki do you think we should do this here or in a followup PR?

**Comment by Czaki** [comment]
> If there isn't an easy way to avoid building the docs twice I vote we push forwards on fixing links - @Czaki do you think we should do this here or in a followup PR?

then we need to change call to `make linkcheck-files || true` to not block other PR with failing CI checks. 

> @Czaki and @laysauchoa do you think there is a way to avoid building the docs twice? It's not obvious to me, unless we had some mechanism to just resolve all links in the docs in advance.

I do not know. Initial Issue suggest to use another tool: [linkchecker](https://github.com/linkchecker/linkchecker)

**Comment by alisterburt** [comment]
> then we need to change call to make linkcheck-files || true to not block other PR with failing CI checks.

okay let's fix the broken links in this PR and merge in one fell swoop

> I do not know. Initial Issue suggest to use another tool: [linkchecker](https://github.com/linkchecker/linkchecker)

I think using the sphinx build tool is fine since we're already using it elsewhere in the docs workflow/makefile - will play with running it before building docs

**Comment by andy-sweet** [comment]
@laysauchoa : just checking on this PR as a candidate for the upcoming v0.4.17 release. We're planning to make the first release candidate in a few days (ideally September 12 or 13), and will hopefully settle on a final candidate over the next week or so.

As this should only be updating the docs and build/CI tools, there's no rush to get this in for the first release candidate it can come in a later one. Do you think you could get the remaining work done by then?

At a minimum, I think we should at least get the the link fixes in, as napari.org redirects to the docs of the latest release. The build/CI tooling could come after the release if needed.

**Comment by Czaki** [comment]
When iterating over links to fix, I found such problems:
```
(naps/2-conda-based-packaging: line  600) broken    https://github.com/napari/napari/blob/be43e127a079f999d830c457d6d69b7c2b56875d/napari/plugins/__init__.py#L34-L42 - Anchor 'L34-L42' not found
```
or
```
(plugins/test_deploy: line  135) broken    https://github.com/napari/cookiecutter-napari-plugin#set-up-automatic-deployments - Anchor 'set-up-automatic-deployments' not found
```
That are valid anchors in the browser, but If I correctly understand, the JS is used to scroll to it. There is a need to check if there is an option to add ignore pattern.

**Comment by Czaki** [comment]
there is a big problem with getting started tutorial, as we have versioned docs, and it is impossible to generate a proper link before making a release

https://napari.org/stable/tutorials/fundamentals/quick_start.html#installation

**Comment by andy-sweet** [comment]
> there is a big problem with getting started tutorial, as we have versioned docs, and it is impossible to generate a proper link before making a release
> 
> https://napari.org/stable/tutorials/fundamentals/quick_start.html#installation

Great point. The URLs are predictable, but if we try to push the predicted URL just before making the RC this check will fail. An ideal solution would be to have some kind of redirect or alias for the latest releases, but unsure if that's possible with GitHub.

**Comment by Czaki** [comment]
> The URLs are predictable, but if we try to push the predicted URL just before making the RC

There is an option to ignore URLs matching given regexps, so we could workaround it by price of do not check the validity of links. 

The other problem here is that someone needs to remember to update it before every release.

File: .github/workflows/build_docs.yml [hunk 2]
```diff
@@ -39,6 +39,9 @@ jobs:
         with:
           run: make docs
 
+      - name: LinkCheck
+        run: make linkcheck-files
```
Comment (andy-sweet) [inline .github/workflows/build_docs.yml]:
```suggestion
      # TODO: enable when the link check passes
      # - name: LinkCheck
      #  run: make linkcheck-files
```

File: .github/workflows/build_docs.yml [hunk 3]
```diff
@@ -38,6 +38,9 @@ jobs:
           GOOGLE_CALENDAR_API_KEY: ${{ secrets.GOOGLE_CALENDAR_API_KEY }}
         with:
           run: make docs
+      # TODO: enable when the link check passes
```
Comment (andy-sweet) [inline .github/workflows/build_docs.yml:42]:
```suggestion

      # TODO: enable when the link check passes
```

**Review by andy-sweet — APPROVED** [review]
I manually checked a lot of these URL changes, and all of them looked good, though some were just redirect changes. I disabled the workflow step, but left all the other code so that we can iterate on this.

**Comment by andy-sweet** [comment]
I will merge this very soon, so that we can get this in for v0.4.17.

**Review by Czaki — APPROVED** [review]

**Comment by andy-sweet** [comment]
Thanks for the work here @laysauchoa and @Czaki !

**Pull request merged**